runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,320 @@
+local util = require('vim.lsp.util')
+local log = require('vim.lsp.log')
+local ms = require('vim.lsp.protocol').Methods
+local api = vim.api
+
+local M = {}
+
+---@class (private) vim.lsp.folding_range.BufState
+---
+---@field version? integer
+---
+--- Index in the form of row -> foldlevel
+---@field row_level table<integer, [integer, nil | ">" | "<"]>
+---
+--- Index in the form of start_row -> collapsed_text
+---@field row_text table<integer, string>
+---
+--- Index in the form of kind -> range
+---@field kind_ranges table<lsp.FoldingRangeKind, lsp.FoldingRange[]?>>
+
+---@type table<integer, vim.lsp.folding_range.BufState?>
+local bufstates = {}
+
+--- Update visible folds in the buffer.
+---@param bufnr integer
+local function foldupdate(bufnr)
+  for _, winid in ipairs(vim.fn.win_findbuf(bufnr)) do
+    local wininfo = vim.fn.getwininfo(winid)[1]
+    if wininfo and wininfo.tabnr == vim.fn.tabpagenr() then
+      if vim.wo[winid].foldmethod == 'expr' then
+        vim._foldupdate(winid, 0, api.nvim_buf_line_count(bufnr))
+      end
+    end
+  end
+end
+
+local augroup_update = api.nvim_create_augroup('vim_lsp_folding_range/update', {})
+
+--- Schedule `foldupdate()` after leaving insert mode.
+---@param bufnr integer
+local function schedule_foldupdate(bufnr)
+  if #(api.nvim_get_autocmds({
+    group = augroup_update,
+    buffer = bufnr,
+  })) == 0 then
+    api.nvim_create_autocmd('InsertLeave', {
+      group = augroup_update,
+      buffer = bufnr,
+      once = true,
+      callback = function()
+        foldupdate(bufnr)
+      end,
+    })
+  end
+end
+
+--- |lsp-handler| for the method `textDocument/foldingRange`
+--- Store folding range then update folds for a specific buffer.
+---@private
+---@param result lsp.FoldingRange[]?
+---@param ctx lsp.HandlerContext
+function M.on_folding_range(err, result, ctx)
```

Comment:
Should this collect the results of all servers and process them in one loop so we then only schedule one update at the end?

Reply:
I am currently not considering support for multiple servers. This is because such functionality is generally provided by a single server, and I feel that vim's own fold feature is unlikely to support this (correct me if I'm wrong), as multiple sources of `foldingRange` may have intersections, making it impossible to calculate the `foldlevel` in that case.

Reply:
Supporting multiple providers is valuable as it opens the door for plugins to implement custom folding via the LSP interface.

Supporting multiple servers is something we've put a lot of effort into recently and want to support it for every method.

Looking for intersections would be a problem, but this implementation just needs to make sure they are handled gracefully, and not necessarily correctly. I think the fold backend will already do something graceful so you may not need to do anything here, we just need to make sure the plumping is in place.

Reply:
I can make some more attempts in this area, but I hope that will be in another PR. Currently (only supporting one server), it can handle most needs. Even implementing custom folding through the LSP interface often covers situations that LSP server does not support folding on a specific filetype. I anticipate that such functionality may bring a lot of complexity, and it would be better to make modifications when support for a single server is already mature, as this proves that providing LSP folding through `foldexpr` is feasible.

Reply:
Then this PR at the very least needs to make sure that multiple clients are handled gracefully by not processing multiple responses and probably only make a request to a single server.

Reply:
After weighing the options, considering that this PR already includes some code for handling multiple clients, I plan to add preliminary support within this PR. This might be better than asking the user to choose which client to use when encountering multiple clients. However, this shouldn’t be the main goal, especially since I can’t find any use case in practical scenarios that would allow me to perform manual testing.